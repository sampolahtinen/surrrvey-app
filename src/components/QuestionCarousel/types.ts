import { APISurveyQuestion } from "../../api/types";

export interface QuestionCarouselProps {
  questions: APISurveyQuestion[];
  onOptionSelect: (param: string) => void;
  selectedOption: string;
  onQuestionChange: any;
}
