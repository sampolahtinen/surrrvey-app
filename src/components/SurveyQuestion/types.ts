export interface SurveyQuestionProps {
  title: string;
  options: string[];
  onClick: (param: string) => void;
  selectedOption: string;
}
