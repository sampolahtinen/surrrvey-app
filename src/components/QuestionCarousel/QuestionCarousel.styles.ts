import styled, { css } from "styled-components";

import { Button } from "../Elements/Button";

export const ButtonsWrapper = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const NextQuestionButton = styled(Button)`
  ${props =>
    props.disabled &&
    css`
      background-color: grey;
      border: 0;
    `}
`;

export const QuestionCarouselContainer = styled.div`
  width: 100%;
`;

export const PrevQuestionButton = styled(NextQuestionButton)``;
