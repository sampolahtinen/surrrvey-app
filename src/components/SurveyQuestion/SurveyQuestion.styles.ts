import styled from "styled-components";
import { colors } from "../../styles/colors";

export const QuestionList = styled.ul``;

export const QuestionTitle = styled.h3`
  margin-bottom: 2.4rem;
`;

export const QuestionsContainer = styled.div`
  width: 400px;
  height: 300px;
`;

export const Option = styled.li<{ selected: boolean }>`
  margin-bottom: 1.6rem;
  border-radius: 12px;
  background-color: ${props =>
    props.selected ? colors.darkLila : colors.offWhite};
  color: ${props => (props.selected ? colors.offWhite : colors.darkGrey)};
  padding: 0.8rem;
  &:hover {
    background-color: ${colors.darkLila};
    color: ${colors.offWhite};
  }
`;
