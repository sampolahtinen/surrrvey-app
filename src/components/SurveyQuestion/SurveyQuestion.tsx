import React from "react";
import {
  QuestionsContainer,
  QuestionTitle,
  QuestionList,
  Option
} from "./SurveyQuestion.styles";
import { SurveyQuestionProps } from "./types";

const SurveyQuestion: React.FC<SurveyQuestionProps> = ({
  options,
  selectedOption,
  title,
  onOptionSelect
}) => {
  return (
    <QuestionsContainer style={{ marginBottom: "3.2rem" }}>
      <QuestionTitle>{title}</QuestionTitle>
      <QuestionList>
        {options.map((option: string) => (
          <Option
            key={title + option}
            onClick={() => onOptionSelect(option)}
            selected={selectedOption === option}
          >
            {option}
          </Option>
        ))}
      </QuestionList>
    </QuestionsContainer>
  );
};

export default SurveyQuestion;
