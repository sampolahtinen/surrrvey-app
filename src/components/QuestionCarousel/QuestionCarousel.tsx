import React, { FC, useState, useEffect } from "react";
import SurveyQuestion from "../SurveyQuestion";
import styled, { css } from "styled-components";
import { Button } from "../Elements/Button";
import { QuestionCarouselProps } from "./types";

export const ButtonsWrapper = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const NextQuestionButton = styled(Button)`
  ${props =>
    props.disabled &&
    css`
      background-color: grey;
      border: 0;
    `}
`;

export const PrevQuestionButton = styled(NextQuestionButton)``;

const QuestionCarousel: FC<QuestionCarouselProps> = ({
  questions,
  onOptionSelect,
  selectedOption,
  onQuestionChange
}) => {
  const [currentQuestionIndex, setCurrentQuestion] = useState(0);

  useEffect(() => {
    onQuestionChange(currentQuestionIndex);
  }, [currentQuestionIndex, onQuestionChange]);

  const handleNext = () => setCurrentQuestion(currentQuestionIndex + 1);
  const handlePrev = () => setCurrentQuestion(currentQuestionIndex - 1);

  return (
    <div>
      <SurveyQuestion
        title={questions[currentQuestionIndex].title}
        options={questions[currentQuestionIndex].options}
        selectedOption={selectedOption}
        onOptionSelect={onOptionSelect}
      />
      <ButtonsWrapper>
        <PrevQuestionButton
          onClick={handlePrev}
          disabled={currentQuestionIndex === 0}
        >
          Back
        </PrevQuestionButton>
        <NextQuestionButton
          onClick={handleNext}
          disabled={currentQuestionIndex === questions.length - 1}
        >
          Next
        </NextQuestionButton>
      </ButtonsWrapper>
    </div>
  );
};

export default QuestionCarousel;
