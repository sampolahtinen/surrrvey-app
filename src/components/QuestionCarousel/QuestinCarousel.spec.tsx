import React from "react";
import { shallow, mount } from "enzyme";
import QuestionCarousel from "./QuestionCarousel";
import SurveyQuestion from "../SurveyQuestion";
import {
  NextQuestionButton,
  PrevQuestionButton
} from "./QuestionCarousel.styles";

const mockQuestions = [
  {
    id: "001",
    title: "firstQuestion",
    options: ["test1", "test2"]
  },
  {
    id: "002",
    title: "secondQuestion",
    options: ["test1", "test2"]
  }
];

describe("QuestionCarousel Component", () => {
  const mockFunc = jest.fn();
  const anotherMockFunc = jest.fn();
  const wrapper = mount(
    <QuestionCarousel
      questions={mockQuestions}
      onOptionSelect={mockFunc}
      selectedOption={"test1"}
      onQuestionChange={anotherMockFunc}
    />
  );

  it("renders", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("switches question when user clicks Next", () => {
    wrapper.find(NextQuestionButton).simulate("click");
    expect(wrapper.find(SurveyQuestion).prop("title")).toEqual(
      "secondQuestion"
    );
  });

  it("triggers onQuestionChange when currentQuestionIndex changes", () => {
    expect(anotherMockFunc).toHaveBeenCalled();
  });

  it("disables next button when displaying the last element", () => {
    expect(wrapper.find(NextQuestionButton).prop("disabled")).toBeTruthy();
  });

  it("can show previous question when user clicks Back", () => {
    wrapper.find(PrevQuestionButton).simulate("click");
    expect(wrapper.find(SurveyQuestion).prop("title")).toEqual("firstQuestion");
  });

  it("disables Back button when displaying the first element", () => {
    expect(wrapper.find(PrevQuestionButton).prop("disabled")).toBeTruthy();
  });
});

// Need to test styled component, because its style is adapted based on props
describe("Next- and PrevQuestionButton Component", () => {
  it("renders properly", () => {
    const wrapper = shallow(<NextQuestionButton />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly when disabled", () => {
    const wrapper = shallow(<NextQuestionButton disabled />);
    expect(wrapper).toMatchSnapshot();
  });
});
