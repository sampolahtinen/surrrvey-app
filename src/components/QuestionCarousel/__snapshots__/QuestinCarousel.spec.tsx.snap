// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Next- and PrevQuestionButton Component renders correctly when disabled 1`] = `
<button
  className="sc-AykKG sc-AykKI hzYser"
  disabled={true}
/>
`;

exports[`Next- and PrevQuestionButton Component renders properly 1`] = `
<button
  className="sc-AykKG sc-AykKI bzjxMr"
/>
`;

exports[`QuestionCarousel Component renders 1`] = `
<QuestionCarousel
  onOptionSelect={[MockFunction]}
  onQuestionChange={
    [MockFunction] {
      "calls": Array [
        Array [
          0,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  questions={
    Array [
      Object {
        "id": "001",
        "options": Array [
          "test1",
          "test2",
        ],
        "title": "firstQuestion",
      },
      Object {
        "id": "002",
        "options": Array [
          "test1",
          "test2",
        ],
        "title": "secondQuestion",
      },
    ]
  }
  selectedOption="test1"
>
  <styled.div
    className="question-carousel"
  >
    <div
      className="sc-AykKJ OwdgA question-carousel"
    >
      <SurveyQuestion
        onOptionSelect={[MockFunction]}
        options={
          Array [
            "test1",
            "test2",
          ]
        }
        selectedOption="test1"
        title="firstQuestion"
      >
        <styled.div
          style={
            Object {
              "marginBottom": "3.2rem",
            }
          }
        >
          <div
            className="sc-AykKE gNSoTB"
            style={
              Object {
                "marginBottom": "3.2rem",
              }
            }
          >
            <styled.h3>
              <h3
                className="sc-AykKD bLtnPA"
              >
                firstQuestion
              </h3>
            </styled.h3>
            <styled.ul>
              <ul
                className="sc-AykKC fmcvyS"
              >
                <styled.li
                  key="firstQuestiontest1"
                  onClick={[Function]}
                  selected={true}
                >
                  <li
                    className="sc-AykKF jqoTha"
                    onClick={[Function]}
                    selected={true}
                  >
                    test1
                  </li>
                </styled.li>
                <styled.li
                  key="firstQuestiontest2"
                  onClick={[Function]}
                  selected={false}
                >
                  <li
                    className="sc-AykKF iPtPKR"
                    onClick={[Function]}
                    selected={false}
                  >
                    test2
                  </li>
                </styled.li>
              </ul>
            </styled.ul>
          </div>
        </styled.div>
      </SurveyQuestion>
      <styled.div>
        <div
          className="sc-AykKH iijtjJ"
        >
          <Styled(Styled(styled.button))
            disabled={true}
            onClick={[Function]}
          >
            <button
              className="sc-AykKG sc-AykKI sc-AykKK gOdPHc"
              disabled={true}
              onClick={[Function]}
            >
              Back
            </button>
          </Styled(Styled(styled.button))>
          <Styled(styled.button)
            disabled={false}
            onClick={[Function]}
          >
            <button
              className="sc-AykKG sc-AykKI bzjxMr"
              disabled={false}
              onClick={[Function]}
            >
              Next
            </button>
          </Styled(styled.button)>
        </div>
      </styled.div>
    </div>
  </styled.div>
</QuestionCarousel>
`;
